File: Calto-GG/Microsoft/Day 2/Rotate_Function.py
class Solution(object):
    def maxRotateFunction(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # initialize the maximum value of F(k) to be the initial value of F(0)
        max_val = sum(i * nums[i] for i in range(len(nums)))
        # initialize the sum of all elements in nums
        total_sum = sum(nums)
        # initialize the current value of F(k) to be the initial value of F(0)
        curr_val = max_val
        # iterate over the possible values of k
        for k in range(1, len(nums)):
            # update the current value of F(k)
            curr_val = curr_val + total_sum - len(nums) * nums[-k]
            # update the maximum value of F(k) if necessary
            max_val = max(max_val, curr_val)
        return max_val

File: nilay302/Microsoft/Day 2/Rotate_function.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int sum = 0;
        int f0 = 0;
        for(int i = 0;i<nums.size();i++){
            sum+=nums[i];
            f0 += i*nums[i];
        }
        vector<int>dp(nums.size(),0);
        dp[0] = f0;
        int max_ans = dp[0];
        int n = nums.size();
        for(int i = 1;i<nums.size();i++){
            dp[i] = dp[i-1] + sum - n*nums[n-i];
            max_ans = max(dp[i],max_ans);
        }
        return max_ans;
    }
};